jax/_src/scipy/stats/bootstrap.py

```diff
@@ -0,0 +1,240 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import NamedTuple
+
+import scipy.stats
+import textwrap
+
+import jax
+from jax import random, vmap
+import jax.numpy as jnp
+from jax._src.scipy.special import ndtr, ndtri
+from jax._src.random import _check_prng_key
+from jax._src.numpy.util import _wraps
+
+_replace_random_state_by_key_doc = textwrap.dedent("""
+Does not support the Scipy argument ``random_state=None``.
+Instead we add an argument ``key`` which is a :class:`jax.random.PRNGKey` random key.
+"""
+)
+_replace_random_state_by_key_no_batch_jnp_statistic_doc = (
+  _replace_random_state_by_key_doc
+  + "\nDoes not support the Scipy argument ``batch``."
+    "\nIn addition, statistic is suggested to implement in JAX. Numpy implementation may lead to error."
+)
+
+
+class ConfidenceInterval(NamedTuple):
+  low: jnp.ndarray
+  high: jnp.ndarray
+
+
+class BootstrapResult(NamedTuple):
+  confidence_interval: ConfidenceInterval
+  standard_error: jnp.ndarray
+
+
+def _bootstrap_resample_and_compute_statistic(sample, statistic, n_resamples, key):
```

Comment:
This would be more consistent with JAX's typical approach if (1) `key` were the first argument, and (2) batching were handled via `vmap` at the call-site, rather than via an argument passed to the function. So the API would be something like this:
```python
def _bootstrap_resample_and_compute_statistic(key, sample, statistic):
  ...
```
and rather than calling it like
```python
_bootstrap_resample_and_compute_statistic(sample, statistic, n_resamples, key)
```
you could instead call it like
```python
keys = random.split(key, n_resamples)
vmap(_bootstrap_resample_and_compute_statistic, (0, None, None))(keys, sample, statistic)
```
The benefits would be (1) more explicit handling of key splitting by the user of the function, and (2) `vmap` at the outer levels may be somewhat more efficient (I'm not entirely sure on that, but I think it is the case) and (3) it's more maintainable, because it makes use of JAX's composable transforms in the way they're intended to be used, rather than hiding them behind less flexible batch-aware APIs.